File: windows/Gutenkern.Core/KerningMarks.cs
namespace Gutenkern;

public enum PairMark
{
    Empty,
    Done
}

public enum GroupMark
{
    Empty,
    Done,
    MixedDone
}

public sealed class KerningMarkState
{
    public HashSet<string> Done { get; }

    public KerningMarkState(IEnumerable<string> done)
    {
        Done = [.. done];
    }

    public PairMark PairMark(string key) =>
        Done.Contains(key) ? Gutenkern.PairMark.Done : Gutenkern.PairMark.Empty;

    public GroupMark GroupMark(IReadOnlyList<string> keys)
    {
        if (keys.Count == 0)
        {
            return Gutenkern.GroupMark.Empty;
        }

        var doneCount = 0;
        foreach (var key in keys)
        {
            if (PairMark(key) == Gutenkern.PairMark.Done)
            {
                doneCount++;
            }
        }

        if (doneCount == keys.Count)
        {
            return Gutenkern.GroupMark.Done;
        }

        if (doneCount > 0)
        {
            return Gutenkern.GroupMark.MixedDone;
        }

        return Gutenkern.GroupMark.Empty;
    }
}

public static class KerningMarks
{
    public static KerningMarkState MarkDone(IReadOnlyList<string> keys, KerningMarkState state)
    {
        var next = Copy(state);
        foreach (var key in keys)
        {
            next.Done.Add(key);
        }

        return next;
    }

    private static KerningMarkState Copy(KerningMarkState state) => new(state.Done);
}

File: macos/Gutenkern/Sources/GutenkernCore/KerningMarks.swift
import Foundation

public enum PairMark: Equatable, Sendable {
    case empty
    case done
}

public enum GroupMark: Equatable, Sendable {
    case empty
    case done
    case mixedDone
}

public struct KerningMarkState: Equatable, Sendable {
    public var done: Set<String>

    public init(done: Set<String> = []) {
        self.done = done
    }

    public func pairMark(_ key: String) -> PairMark {
        done.contains(key) ? .done : .empty
    }

    public func groupMark(keys: [String]) -> GroupMark {
        guard !keys.isEmpty else {
            return .empty
        }
        let doneCount = keys.reduce(into: 0) { count, key in
            if pairMark(key) == .done {
                count += 1
            }
        }
        if doneCount == keys.count {
            return .done
        }
        if doneCount > 0 {
            return .mixedDone
        }
        return .empty
    }
}

public enum KerningMarks {
    public static func markDone(keys: [String], state: KerningMarkState) -> KerningMarkState {
        var next = state
        next.done.formUnion(keys)
        return next
    }
}
